src/libstd/sys/unix/thread.rs

```diff
@@ -125,16 +125,33 @@ impl Thread {
     }
 
     pub fn sleep(dur: Duration) {
-        let mut ts = libc::timespec {
-            tv_sec: dur.as_secs() as libc::time_t,
-            tv_nsec: dur.subsec_nanos() as libc::c_long,
-        };
+        fn get_secs(secs: &mut u64) -> libc::time_t {
+            if *secs > libc::time_t::max_value() as u64 {
+                *secs -= libc::time_t::max_value() as u64;
+                libc::time_t::max_value()
+            } else {
+                let copy = *secs;
+                *secs = 0;
+                copy as libc::time_t
+            }
+        }
+
+        let mut secs = dur.as_secs();
+        let mut nsecs = dur.subsec_nanos() as libc::c_long;
 
         // If we're awoken with a signal then the return value will be -1 and
         // nanosleep will fill in `ts` with the remaining time.
         unsafe {
-            while libc::nanosleep(&ts, &mut ts) == -1 {
-                assert_eq!(os::errno(), libc::EINTR);
+            while secs > 0 || nsecs > 0 {
+                let mut ts = libc::timespec {
+                    tv_sec: get_secs(&mut secs),
+                    tv_nsec: nsecs,
+                };
+                if libc::nanosleep(&ts, &mut ts) == -1 {
+                    assert_eq!(os::errno(), libc::EINTR);
+                }
+                secs += ts.tv_sec as u64;
+                nsecs = ts.tv_nsec;
```

Comment:
Interestingly, documentation for nanosleep does not specify what `ts` contains after a successful call, therefore, in order to be on the safe side, you want to only do this in case the nanosleep call fails and set these fields to 0 otherwise.